File: javascript/binary-search-tree/binary-search-tree.js
export default class BinarySearchTree {
    constructor(value) {
        this._data = value;
        this._left = null;
        this._right = null;
    } 

    get data() {
        return this._data;
    }

    get left() {
        return this._left;
    }

    get right() {
        return this._right;
    }

    insert(value) {
        if (value <= this.data)
            this._left = this.insertIntoTree(value, this._left);
        else
            this._right = this.insertIntoTree(value, this._right);
    }

    insertIntoTree(value, tree) {
        if (tree === null)
            return new BinarySearchTree(value);

        tree.insert(value);
        return tree;
    }

    each(callback) {
        if (this.left)
            this.left.each(callback);

        callback(this.data);

        if (this.right)
            this.right.each(callback);
    }
}


//     private static BinarySearchTree Add(int value, BinarySearchTree tree)
//     {
//         if (tree == null)
//         {
//             return new BinarySearchTree(value);
//         }

//         return tree.Add(value);
//     }

//     public IEnumerator<int> GetEnumerator()
//     {
//         foreach (var left in Left?.AsEnumerable() ?? Enumerable.Empty<int>())
//         {
//             yield return left;
//         }

//         yield return Value;

//         foreach (var right in Right?.AsEnumerable() ?? Enumerable.Empty<int>())
//         {
//             yield return right;
//         }
//     }

//     IEnumerator IEnumerable.GetEnumerator()
//     {
//         return GetEnumerator();
//     }
// }

File: csharp/binary-search-tree/BinarySearchTree.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

public class BinarySearchTree : IEnumerable<int>
{
    public BinarySearchTree(params int[] values)
    {
        if (values.Length == 0) throw new ArgumentException("Cannot create tree from empty list");

        Value = values[0];

        foreach (var value in values.Skip(1))
            Add(value);
    }

    public int Value { get; }

    public BinarySearchTree Left { get; private set; }

    public BinarySearchTree Right { get; private set; }

    private BinarySearchTree Add(int value)
    {
        if (value <= Value)
            Left = Add(value, Left);
        else
            Right = Add(value, Right);

        return this;
    }

    private static BinarySearchTree Add(int value, BinarySearchTree tree)
    {
        if (tree == null)
            return new BinarySearchTree(value);

        return tree.Add(value);
    }

    public IEnumerator<int> GetEnumerator()
    {
        foreach (var left in Left?.AsEnumerable() ?? Enumerable.Empty<int>())
            yield return left;

        yield return Value;

        foreach (var right in Right?.AsEnumerable() ?? Enumerable.Empty<int>())
            yield return right;
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}